File: java/BinaryHeap.java
// complete binary tree, meaning every level is filled except possibly the last which is filled from left 
// to right
// max heap; value of each node >= values of children
// min heap; value of each node <= values of children 

import java.util.ArrayList;
import java.util.List;

public class BinaryHeap {
  private final List<Integer> heap;
  private final boolean isMaxHeap;

  public BinaryHeap(boolean isMaxHeap) {
    this.heap = new ArrayList<>();
    this.isMaxHeap = isMaxHeap;
  }
  private int getParentIndex(int childIndex) {
    return (childIndex - 1) / 2;
  }
  private int getLeftChildIndex(int parentIndex) {
    return 2 * parentIndex + 1;
  }
  private int getRightChildIndex(int parentIndex) {
    return 2 * parentIndex + 2;
  }
  private void heapifyUp(int index) {
    int parentIndex = getParentIndex(index);
    if(parentIndex >= 0) {
      if(isMaxHeap) {
        if(heap.get(index) > heap.get(parentIndex)) {
          swap(index, parentIndex);
          heapifyUp(parentIndex);
        }
      } else {
        if(heap.get(index) < heap.get(parentIndex)) {
          swap(index, parentIndex);
          heapifyUp(parentIndex);
        }
      }
    }
  }
  private void heapifyDown(int index) {
    int leftChildIndex = getLeftChildIndex(index);
    int rightChildIndex = getRightChildIndex(index);
    int largestOrSmallestIndex = index;

    if(isMaxHeap) {
      if(leftChildIndex < heap.size() && heap.get(leftChildIndex) > heap.get(largestOrSmallestIndex)) {
        largestOrSmallestIndex = leftChildIndex;
      }
      if(rightChildIndex < heap.size() && heap.get(rightChildIndex) > heap.get(largestOrSmallestIndex)){
        largestOrSmallestIndex = rightChildIndex;
      }
    } else {
      if(leftChildIndex < heap.size() && heap.get(leftChildIndex) < heap.get(largestOrSmallestIndex)) {
        largestOrSmallestIndex = leftChildIndex;
      }
      if(rightChildIndex < heap.size() && heap.get(rightChildIndex) < heap.get(largestOrSmallestIndex)) {
        largestOrSmallestIndex = rightChildIndex;
      }
    }
    if(largestOrSmallestIndex != index) {
      swap(index, largestOrSmallestIndex);
      heapifyDown(largestOrSmallestIndex);
    }
  }

  private void swap(int index1, int index2) {
    int temp = heap.get(index1);
    heap.set(index1, heap.get(index2));
    heap.set(index2, temp);
  }
  public void insert(int value) {
    heap.add(value);
    heapifyUp(heap.size() - 1);
  }
  public int extractMax() {
    if(heap.isEmpty()) {
      throw new IllegalStateException("heap is empty");
    }
    int max = heap.get(0);
    swap(0, heap.size() - 1);
    heap.remove(heap.size() - 1);
    heapifyDown(0);
    return max;
  }

  public int extractMin() {
    if(heap.isEmpty()) {
      throw new IllegalStateException("Heap is empty");
    }
    int min = heap.get(0);
    swap(0, heap.size() - 1);
    heap.remove(heap.size() - 1);
    heapifyDown(0);
    return min;
  }

  public int peek() {
    if(heap.isEmpty()) {
      throw new IllegalStateException("heap is empty");
    }
    return heap.get(0);
  }

  public boolean isEmpty() {
    return heap.isEmpty();
  }
}
    

File: python/binary_heap.py
# operations and their time complexities
# insert; O(log n)
# extract min/max; O(log n)
# heapify up/down; O(log n) 



class BinaryHeap:
  def __init__(self, heap_type='min'):
    self.heap = []
    self.heap_type = heap_type

  def parent(self, i):
    return (i - 1) // 2

  def left_child(self, i):
    return 2 * i + 1

  def right_child(self, i):
    return 2 * i + 2

  def heapify_up(self, i):
    while i > 0 and self.heap[self.parent(i)] > self.heap[i] if self.heap_type == 'min' else self.heap[self.parent(i)] < self.heap[i]:
      self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
      i = self.parent(i)

  def heapify_down(self, i):
    while True:
      largest = i
      left = self.left_child(i)
      right = self.right_child(i)

      if left < len(self.heap) and self.heap[left] < self.heap[largest] if self.heap_type == 'min' else self.heap[left] > self.heap[largest]:
        largest = left

      if right < len(self.heap) and self.heap[right] < self.heap[largest] if self.heap_type == 'min' else self.heap[right] > self.heap[largest]:
        largest = right 

      if largest != i:
        self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
        i = largest

      else:
        break

    def insert(self, value):
      self.heap.append(value)
      self.heapify_up(len(self.heap) - 1)

    def extract_min(self):
      if not self.heap:
        return None
      if len(self.heap) == 1:
        return self.heap.pop()
      root = self.heap[0]
      self.heap[0] = self.heap.pop()
      return root

    def extract_max(self):
      if not self.heap:
        return None
      if len(self.heap) == 1:
        return self.heap..pop()
      root = self.heap[0]
      self.heap[0] = self.heap.pop()
      self.heapify_down(0)
      return root 

